components/user-mgt/org.wso2.carbon.user.mgt/src/main/java/org/wso2/carbon/user/mgt/UserAdmin.java

```diff
@@ -280,19 +281,26 @@ private void addUserRole(String roleName, String[] userList, String[] permission
         if (!isAllowedRoleName(roleName, realm)) {
             throw new UserAdminException("Role name is reserved by the system.");
         }
-        if (!isInternalRole) {
-            if (getUserAdminProxy().isRoleAndGroupSeparationEnabled()) {
-                if (ArrayUtils.isNotEmpty(permissions)) {
-                    addInternalSystemRole(roleName, permissions);
-                    permissions = new String[0];
-                }
-            }
+        if (isInternalRole) {
+            getUserAdminProxy().addInternalRole(roleName, userList, permissions);
+        } else if (!(getUserAdminProxy().isRoleAndGroupSeparationEnabled()) || ArrayUtils.isEmpty(permissions)) {
```

Comment:
why do we check this ArrayUtils.isEmpty(permissions)?

Are we then creating the system role while updating the role again with permissions?

Reply:
If the permission is empty then we are not creating the system role, when updating the role first we will check whether a relevant internal role exists, and then only we are creating the system role.